### DISCUSSION TITLE
tsc runs duplicate work

### DISCUSSION THREAD

**Discussion by connorjs** [vercel/turborepo#4143] [Help] [answered]
I have followed the [TypeScript guide](https://turbo.build/repo/docs/handbook/linting/typescript) and read [the "main" TypeScript blog post](https://turbo.build/blog/you-might-not-need-typescript-project-references) by Jared. (I referenced the original one a couple years back from tsdx days.)

Now that I'm using turbo repo though, I'm uncertain how to handle `tsc`.

**First attempt**: Use "internal" TypeScript packages and include `tsc` in each package. We want `tsc` in each package to speed up builds and turborepo handles the orchestration and caching.

However, I _think_ this results in duplicate work. If we have A depends on B, then B gets type-checked twice: Once during B's `tsc` and again during A's `tsc` due to the nature of "internal" TypeScript packages. _I think_.

~~**Can someone confirm this?** (I used `--explainFiles` option of `tsc` and B's contents in A's run.)~~

Edit: I confirmed this by intentionally inserting faulty TS code and running `turbo run tsc --continue` ([continue docs](https://turbo.build/repo/docs/reference/command-line-reference#--continue)).

**Second (future) attempt**: Therefore, it seems like we have two choices: (1) Only run `tsc` at the application level (which still has duplicates if packages shared across apps) or (2) go back to TypeScript project references.

Should I try out project references? I have avoided them for a while given the article.

**Comment by dbrxnds** [comment]
Hi @connorjs, wondering what you ended up doing with this

**Comment by wouter-leistra** [comment]
Shouldn't it after it has run for the package itself get a cache hit in each of the apps that list the typescript package as a dependency?

**Reply by davidgoli** [reply]
I don't think this would help if, as the handbook recommends, you're not writing `.d.ts` files. While you'll get a cache hit, the problem remains that it doesn't save you anything - your root module will still have to transpile everything it imports.

**Comment by davidgoli** [comment]
My reading of the guide and the blogpost hasn't lead me to a clear understanding of the benefits of turborepo for TypeScript monorepos. As the blogpost says:

> The consuming application of an internal package must transpile and typecheck it.

If this is true, then the entire monorepo must be transpiled in a single pass, and there is no possibility to parallelize the task of typechecking - the TypeScript compiler must traverse the entire module graph in order to build the types necessary to check the code in the consuming application.

Thus, it would appear that Turborepo doesn't really help speed up TypeScript compilation using the TypeScript package guidelines their own docs provide.

Alternatively, the blogpost continues:

> When/if overall build time begins to suffer, you might decide to convert your larger internal packages back into "regular" packages with .d.ts files and with normal TypeScript build steps.

If my understanding is incorrect, I would love to learn otherwise!

**Comment by baptisteArno** [comment]
Right, it is not optimal for sure.

I am currently trying to set the exports type field to point to built d.ts files. Along with `skipLibCheck: true` it should simply take the d.ts files as is and won't have to build all its dependencies.

**Comment by anthonyshew** [comment, marked as answer]
Hey, folks, just took a look a this as I ran into it myself and verified the behavior with `create-turbo`.

What I'm finding is that [Just-in-Time Packages](https://turbo.build/repo/docs/core-concepts/internal-packages#just-in-time-packages) end up getting checked by their consumer. This makes sense, since the consumer is being fed raw TypeScript, so it's getting included in that program's files. Therefore, from TypeScript's perspective, those files need type checked.

`create-turbo` uses Just-in-Time Packages, so I switched it the `@repo/ui` package to be a [Compiled Package](https://turbo.build/repo/docs/core-concepts/internal-packages#compiled-packages) to see if there would be a difference. It looks like it does! I purposefully added a type error into the `@repo/ui` package and it doesn't get reported when running a `turbo run type-check --filter=web --only` command.

If anyone here wants to check my understanding of this, please do! We can write this into the tradeoffs section of the page linked above.

**Reply by jfraudeau** [reply]
Hi, I just tried this approach and I do get errors from `@repo/ui` when running `pnpm run check-types --filter web --only --force` when following [Compiled Package](https://turbo.build/repo/docs/core-concepts/internal-packages#compiled-packages) which means keeping the ts sources as the types exports.

This is because `skipLibCheck` only skips checks for `d.ts` files as @baptisteArno indicates. In order to avoid duplicate work it means that we need to put the declaration files as types exports. It also means that `check-types` must depends on `^build` otherwise type declarations are not available, the usefulness of having `check-types` separate from `build` is very slim outside of top level packages in this case.

I'm not sure this solution will be the fastest for everyone but the tradeoffs of exporting `.ts` files could be bettor documented

**Comment by SohelIslamImran** [comment]
This is creating a huge issue if you have a different tsconfig for the dependent package.. When running the task in a dependent package, it uses the tsconfig of the parent package. I have created an issue https://github.com/vercel/turborepo/issues/10863

**Reply by anthonyshew** [reply]
https://github.com/vercel/turborepo/discussions/4143#discussioncomment-10665480
